### PR TITLE
Backend refactor, Performance improvements, Reactions

### PR DESCRIPTION
**Pull request by rawdigits**
This started as a branch to support reactions, but to do so properly required a number of backend changes. In no particular order:

* Enhanced SearchList class and made much more performant. Added a hashtable to each SearchList that acts as an index for much faster lookup of objects when doing `.find()`
* Added `Message` object. When storing a message in `channel.messages`, it should be this type.
* The global `message_cache` is now a list instead of a hash. this allows a simpler format where the message json is just stored directly to disk and replayed via process_message.
* Removed `Meta` class.
* Removed `SlackThing` superclass.
* Added some documentation.

### CODE REVIEW

File: wee_slack.py [hunk 1]
```diff
@@ -60,97 +63,33 @@ def dbg(message, fout=False, main_buffer=False):
...
-
-    def get_all(self, attribute):
-        items = []
-        for child in self:
-            if child.__class__ == self.__class__:
-                items += child.get_all(attribute)
-            else:
-                items += (eval("child." + attribute))
-        return items
+            if "get_aliases" in dir(child):
```
Comment (tfheen) [inline wee_slack.py]: I like `hasattr(child, "get_aliases") and callable(child.get_aliases)` better. `dir` is more of a debugging tool

File: wee_slack.py [hunk 2]
```diff
@@ -363,33 +301,58 @@ def buffer_input_cb(b, buffer, data):
...
-        self.last_read = float(last_read)
         self.last_received = None
+        self.messages = []
+        self.scrolling = False
         if active:
             self.create_buffer()
             self.attach_buffer()
             self.update_nicklist()
             self.set_topic(self.topic)
+            self.create_members_table()
```
Comment (tfheen) [inline wee_slack.py]: Shouldn't this happen before you call `update_nicklist`?

File: wee_slack.py (hunk 1 above)
Comment (rawdigits) [inline wee_slack.py]: ah yep, much better!

File: wee_slack.py (hunk 2 above)
Comment (rawdigits) [inline wee_slack.py]: it should indeed. good catch.

File: wee_slack.py [hunk 3]
```diff
@@ -1602,10 +1673,30 @@ def url_processor_cb(data, command, return_code, out, err):
...
+                cache_file.write("{}\n".format(json.dumps(message.message_json)))
     return w.WEECHAT_RC_OK
 
+def cache_load():
+    global message_cache
+    try:
+        cache_file = open("{}/{}".format(WEECHAT_HOME, CACHE_NAME), 'r')
+        for line in cache_file:
+            message_cache.append(line)
+    except:
```
Comment (tfheen) [inline wee_slack.py]: Consider just catching the file not found exception?

File: wee_slack.py [hunk 4]
```diff
@@ -1683,21 +1774,39 @@ def config_changed_cb(data, option, value):
...
+            channels.find(w.current_buffer()).unset_scrolling()
+    except:
+        pass
+    return w.WEECHAT_RC_OK
+
 # END Utility Methods
 
 # Main
 if __name__ == "__main__":
+
```
Comment (tfheen) [inline wee_slack.py]: Remove those blank lines

**Pull request merged**